File: de.ibapl.jnhw.posix/src/main/java/de/ibapl/jnhw/util/posix/Defines.java
/*
 * JNHW - Java Native header Wrapper, https://github.com/aploese/jnhw/
 * Copyright (C) 2019-2023, Arne Plöse and individual contributors as indicated
 * by the @authors tag. See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
package de.ibapl.jnhw.util.posix;

import de.ibapl.jnhw.common.annotation.Define;
import de.ibapl.jnhw.common.datatypes.BaseDataType;
import de.ibapl.jnhw.common.util.IntDefine;
import de.ibapl.jnhw.libloader.Arch;
import de.ibapl.jnhw.libloader.Endianess;
import de.ibapl.jnhw.libloader.MemoryModel;
import de.ibapl.jnhw.libloader.MultiarchTupelBuilder;
import de.ibapl.jnhw.libloader.OS;

/**
 * get the defines with gcc: create an empty file c.c run
 * {@code gcc -dD -dI -E c.c &gt; c.txt} c.txt contains all macros. add an
 * {@code #include &lt;headerFileName.h>} to get the defines for that header.
 *
 * @author aploese
 */
public class Defines {

    @Define
    public final static IntDefine __aarch64__;

    @Define
    public final static IntDefine __alpha__;

    @Define
    public final static IntDefine __amd64__;

    /**
     */
    @Define
    public final static IntDefine __APPLE__;

    @Define
    public final static IntDefine __arm__;

    @Define
    public final static IntDefine __ARM_ARCH;

    /**
     * The biggest alignemnt used. If sizeof(struct) is bigger than
     * {@code __BIGGEST_ALIGNMENT__} use an address that is a multiple of
     * {@code __BIGGEST_ALIGNMENT__} as {@code baseaddress}.
     *
     * @return
     */
    @Define
    public final static int __BIGGEST_ALIGNMENT__;

    @Define
    public final static int __BYTE_ORDER__;

    /**
     * the major version at compile time
     */
    @Define
    public final static IntDefine __FreeBSD__;

    @Define
    public final static IntDefine __GLIBC__;

    @Define
    public final static IntDefine __GLIBC_MINOR__;

    @Define
    public final static IntDefine __GNU_LIBRARY__;

    @Define
    public final static IntDefine __i386__;

    @Define
    public final static IntDefine __i686__;

    @Define
    public final static IntDefine __ILP32__;

    @Define
    public final static IntDefine __linux__;

    @Define
    public final static IntDefine __LP64__;

    @Define
    public final static IntDefine __mips__;

    @Define
    public final static IntDefine __mips64;

    @Define
    public final static IntDefine __MIPSEB__;

    @Define
    public final static IntDefine __MIPSEL__;

    /**
     *
     */
    @Define
    public final static IntDefine __OpenBSD__;

    @Define
    public final static int __ORDER_BIG_ENDIAN__;

    @Define
    public final static int __ORDER_LITTLE_ENDIAN__;

    @Define
    public final static int __ORDER_PDP_ENDIAN__;

    @Define
    public final static IntDefine __powerpc__;

    @Define
    public final static IntDefine __powerpc64__;

    @Define
    public final static IntDefine __riscv;

    @Define
    public final static IntDefine __s390__;

    @Define
    public final static IntDefine __s390x__;

    @Define
    public final static IntDefine __sh__;

    @Define
    public final static IntDefine __SH4__;

    @Define
    public final static int __SIZEOF_LONG__;

    @Define
    public final static int __SIZEOF_POINTER__;

    @Define
    public final static IntDefine __sparc__;

    @Define
    public final static IntDefine __sparc64__;

    /**
     * size of time_t glibc bits/timesize.h
     *
     */
    @Define
    public final static IntDefine __TIMESIZE;

    /**
     * Its defined at different places: glibc: bits/wordsize.h FreeBSD:
     * sys/stdint.h
     *
     * but not at OpenBSD
     *
     * so we keep his here for the moment. _POSIX_C_SOURCE
     *
     */
    @Define
    public final static IntDefine __WORDSIZE;

    @Define
    public final static IntDefine __x86_64__;

    @Define
    public final static IntDefine _BSD_SOURCE;

    /**
     * maybe defined in jnhw-posix.h
     *
     */
    @Define
    public final static IntDefine _FILE_OFFSET_BITS;

    /**
     * maybe defined in jnhw-posix.h
     *
     */
    @Define
    public final static IntDefine _LARGEFILE_SOURCE;

    /**
     * maybe defined in jnhw-posix.h
     *
     * _LARGEFILE64_SOURCE was defined at native compile time. If
     * _LARGEFILE64_SOURCE was defined then all largefile64 functions (i.e.
     * open64, read64, fseek64, ...) are available.
     *
     *
     * time of the native code.
     *
     */
    @Define
    public final static IntDefine _LARGEFILE64_SOURCE;

    /**
     * maybe defined in jnhw-posix.h
     *
     */
    @Define
    public final static IntDefine _POSIX_C_SOURCE;

    /**
     * maybe defined in jnhw-posix.h
     *
     */
    @Define
    public final static IntDefine _XOPEN_SOURCE;

    /**
     * maybe defined in jnhw-posix.h
     *
     */
    @Define
    public final static IntDefine _XOPEN_SOURCE_EXTENDED;

    /**
     *
     * @implNote The actual value for the define fields are injected by
     * initFields. The static initialization block is used to set the value here
     * to communicate that this static final fields are not statically foldable.
     * {
     * @see String#COMPACT_STRINGS}
     */
    static {

        final Arch arch = MultiarchTupelBuilder.getArch();
        final OS os = MultiarchTupelBuilder.getOS();
        final Endianess e = MultiarchTupelBuilder.getEndianess();

        __aarch64__ = arch == Arch.AARCH64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __alpha__ = IntDefine.UNDEFINED;
        __amd64__ = arch == Arch.X86_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __arm__ = arch == Arch.ARM ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __ARM_ARCH = switch (arch) {
            case AARCH64 ->
                IntDefine.toIntDefine(8);
            case ARM ->
                IntDefine.toIntDefine(7);
            default ->
                IntDefine.UNDEFINED;
        };
        __i386__ = arch == Arch.I386 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __i686__ = arch == Arch.I386 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __mips__ = (arch == Arch.MIPS) || (arch == Arch.MIPS_64) ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __mips64 = arch == Arch.MIPS_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __MIPSEB__ = ((arch == Arch.MIPS) || (arch == Arch.MIPS_64)) && e == Endianess.BIG ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __MIPSEL__ = ((arch == Arch.MIPS) || (arch == Arch.MIPS_64)) && e == Endianess.LITTLE ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __powerpc__ = arch == Arch.POWER_PC_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __powerpc64__ = arch == Arch.POWER_PC_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __riscv = arch == Arch.RISC_V_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __SH4__ = IntDefine.UNDEFINED;
        __s390__ = arch == Arch.S390_X ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __s390x__ = arch == Arch.S390_X ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __sh__ = IntDefine.UNDEFINED;
        __sparc64__ = arch == Arch.SPARC_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __sparc__ = arch == Arch.SPARC_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __x86_64__ = arch == Arch.X86_64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;

        __APPLE__ = os == OS.DARWIN ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        _BSD_SOURCE = os == OS.OPEN_BSD ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __FreeBSD__ = os == OS.FREE_BSD ? IntDefine.toIntDefine(13) : IntDefine.UNDEFINED;
        __linux__ = os == OS.LINUX ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;
        __OpenBSD__ = os == OS.OPEN_BSD ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;

        _FILE_OFFSET_BITS = IntDefine.UNDEFINED;
        switch (os) {
            case DARWIN, FREE_BSD, OPEN_BSD -> {
                _LARGEFILE64_SOURCE = IntDefine.UNDEFINED;
                _LARGEFILE_SOURCE = IntDefine.UNDEFINED;
            }
            case LINUX -> {
                _LARGEFILE64_SOURCE = IntDefine.toIntDefine(1);
                _LARGEFILE_SOURCE = IntDefine.toIntDefine(1);
            }
            case WINDOWS -> {
                _LARGEFILE64_SOURCE = IntDefine.UNDEFINED;
                _LARGEFILE_SOURCE = IntDefine.UNDEFINED;
            }
            default ->
                throw new NoClassDefFoundError("No default value for _LARGEFILE64_SOURCE and _LARGEFILE_SOURCE " + MultiarchTupelBuilder.getMultiarch());
        }
        switch (os) {
            case DARWIN, FREE_BSD, OPEN_BSD, LINUX -> {
                _POSIX_C_SOURCE = IntDefine.toIntDefine(200809);
                _XOPEN_SOURCE = IntDefine.toIntDefine(700);
                _XOPEN_SOURCE_EXTENDED = IntDefine.toIntDefine(1);
            }
            case WINDOWS -> {
                _POSIX_C_SOURCE = IntDefine.UNDEFINED;
                _XOPEN_SOURCE = IntDefine.UNDEFINED;
                _XOPEN_SOURCE_EXTENDED = IntDefine.UNDEFINED;
            }
            default ->
                throw new NoClassDefFoundError("No default value for _POSIX_C_SOURCE,_XOPEN_SOURCE and _XOPEN_SOURCE_EXTENDED " + MultiarchTupelBuilder.getMultiarch());
        }

        switch (arch) {
            case AARCH64, I386, MIPS_64, X86_64, POWER_PC_64, RISC_V_64 ->
                __BIGGEST_ALIGNMENT__ = 16;
            case S390_X ->
                __BIGGEST_ALIGNMENT__ = 8;
            case ARM, MIPS ->
                __BIGGEST_ALIGNMENT__ = 8;
            default ->
                throw new NoClassDefFoundError("No default value for __BIGGEST_ALIGNMENT__ " + MultiarchTupelBuilder.getMultiarch());
        }

        switch (os) {
            case LINUX -> {
                __GLIBC_MINOR__ = IntDefine.toIntDefine(36);
                __GLIBC__ = IntDefine.toIntDefine(2);
                __GNU_LIBRARY__ = IntDefine.toIntDefine(6);
            }
            case DARWIN, FREE_BSD, OPEN_BSD -> {
                __GLIBC_MINOR__ = IntDefine.UNDEFINED;
                __GLIBC__ = IntDefine.UNDEFINED;
                __GNU_LIBRARY__ = IntDefine.UNDEFINED;
            }
            case WINDOWS -> {
                __GLIBC_MINOR__ = IntDefine.UNDEFINED;
                __GLIBC__ = IntDefine.UNDEFINED;
                __GNU_LIBRARY__ = IntDefine.UNDEFINED;
            }
            default ->
                throw new NoClassDefFoundError("No default value for __GLIBC__, __GLIBC_MINOR__ and __GNU_LIBRARY__ " + MultiarchTupelBuilder.getMultiarch());
        }

        __ILP32__ = switch (MultiarchTupelBuilder.getMemoryModel()) {
            case ILP32 ->
                switch (MultiarchTupelBuilder.getMultiarch()) {
                    case MIPS_EL__LINUX__GNU, ARM__LINUX__GNU_EABI_HF ->
                        IntDefine.UNDEFINED;
                    default ->
                        IntDefine.toIntDefine(1);
                };
            default ->
                IntDefine.UNDEFINED;
        };
        __LP64__ = MultiarchTupelBuilder.getMemoryModel() == MemoryModel.LP64 ? IntDefine.toIntDefine(1) : IntDefine.UNDEFINED;

        switch (e) {
            case BIG ->
                __BYTE_ORDER__ = 4321;
            case LITTLE ->
                __BYTE_ORDER__ = 1234;
            default ->
                throw new NoClassDefFoundError("No default value for __BYTE_ORDER__  " + MultiarchTupelBuilder.getMultiarch());
        }
        __ORDER_BIG_ENDIAN__ = 4321;
        __ORDER_LITTLE_ENDIAN__ = 1234;
        __ORDER_PDP_ENDIAN__ = 3412;

        __SIZEOF_LONG__ = MultiarchTupelBuilder.getMemoryModel().sizeOf_long.sizeInByte;
        __SIZEOF_POINTER__ = MultiarchTupelBuilder.getMemoryModel().sizeOf_pointer.sizeInByte;

        // __TIMESIZE = IntDefine.UNDEFINED; // glibc > 2.31? IntDefine.toIntDefine(mi.getSizeOfPointer().sizeInBit);
        //Linux
        __TIMESIZE = switch (os) {
            case LINUX ->
                switch (MultiarchTupelBuilder.getMemoryModel()) {
                    case ILP32 ->
                        IntDefine.toIntDefine(BaseDataType.uint64_t.SIZE_OF * 4);
                    case LP64 ->
                        IntDefine.toIntDefine(BaseDataType.uint64_t.SIZE_OF * 8);
                    default ->
                        throw new RuntimeException("Cant handle memory model: " + MultiarchTupelBuilder.getMemoryModel());
                };
            case WINDOWS ->
                IntDefine.UNDEFINED;
            default ->
                throw new NoClassDefFoundError("No default value for __TIMESIZE  " + MultiarchTupelBuilder.getMultiarch());
        };
        __WORDSIZE = switch (os) {
            case OPEN_BSD, WINDOWS ->
                IntDefine.UNDEFINED;
            default ->
                IntDefine.toIntDefine(MultiarchTupelBuilder.getMemoryModel().sizeOf_pointer.sizeInBit);
        };

    }

}

File: de.ibapl.jnhw.posix/src/test/native/Defines.c
/**
 * JNHW - Java Native header Wrapper, https://github.com/aploese/jnhw/
 * Copyright (C) 2021-2024, Arne Plöse and individual contributors as indicated
 * by the @authors tag. See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
#include "jnhw-posix.h"

#include <stdint.h>
#include <unistd.h>


int32_t* tryGetValueOf__BSD_SOURCE(int32_t* value) {
#if defined (_BSD_SOURCE)
    *value = _BSD_SOURCE;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf__FILE_OFFSET_BITS(int32_t* value) {
#if defined (_FILE_OFFSET_BITS)
    *value = _FILE_OFFSET_BITS;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf__LARGEFILE64_SOURCE(int32_t* value) {
#if defined(_LARGEFILE64_SOURCE)
    *value = _LARGEFILE64_SOURCE;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf__LARGEFILE_SOURCE(int32_t* value) {
#if defined(_LARGEFILE_SOURCE)
    *value = _LARGEFILE_SOURCE;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf__POSIX_C_SOURCE(int32_t* value) {
#if defined (_POSIX_C_SOURCE)
    *value = _POSIX_C_SOURCE;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf__XOPEN_SOURCE(int32_t* value) {
#if defined (_XOPEN_SOURCE)
    *value = _XOPEN_SOURCE;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf__XOPEN_SOURCE_EXTENDED(int32_t* value) {
#if defined (_XOPEN_SOURCE_EXTENDED)
    *value = _XOPEN_SOURCE_EXTENDED;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___aarch64__(int32_t* value) {
#if defined(__aarch64__)
    *value = __aarch64__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___alpha__(int32_t* value) {
#if defined(__alpha__)
    *value = __alpha__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___amd64__(int32_t* value) {
#if defined(__amd64__)
    *value = __amd64__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___APPLE__(int32_t* value) {
#if defined(__APPLE__)
    *value = __APPLE__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___arm__(int32_t* value) {
#if defined(__arm__)
    *value = __arm__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___ARM_ARCH(int32_t* value) {
#if defined(__ARM_ARCH)
    *value = __ARM_ARCH;
#else
    value = NULL;
#endif
    return value;
}

int32_t getValueOf___BIGGEST_ALIGNMENT__() {
    return __BIGGEST_ALIGNMENT__;
}

int32_t getValueOf___BYTE_ORDER__() {
    return __BYTE_ORDER__;
}

int32_t* tryGetValueOf___FreeBSD__(int32_t* value) {
#if defined(__FreeBSD__)
    *value = __FreeBSD__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___GLIBC_MINOR__(int32_t* value) {
#if defined(__GLIBC_MINOR__)
    *value = __GLIBC_MINOR__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___GLIBC__(int32_t* value) {
#if defined(__GLIBC__)
    *value = __GLIBC__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___GNU_LIBRARY__(int32_t* value) {
#if defined(__GNU_LIBRARY__)
    *value = __GNU_LIBRARY__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___i386__(int32_t* value) {
#if defined(__i386__)
    *value = __i386__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___i686__(int32_t* value) {
#if defined(__i686__)
    *value = __i686__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___ILP32__(int32_t* value) {
#if defined (__ILP32__)
    *value = __ILP32__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___linux__(int32_t* value) {
#if defined(__linux__)
    *value = __linux__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___LP64__(int32_t* value) {
#if defined (__LP64__)
    *value = __LP64__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___mips__(int32_t* value) {
#if defined(__mips__)
    *value = __mips__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___mips64(int32_t* value) {
#if defined(__mips64)
    *value = __mips64;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___MIPSEB__(int32_t* value) {
#if defined(__MIPSEB__)
    *value = __MIPSEB__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___MIPSEL__(int32_t* value) {
#if defined(__MIPSEL__)
    *value = __MIPSEL__;
#else
    value = NULL;
#endif
    return value;
}

int32_t getValueOf___ORDER_BIG_ENDIAN__() {
    return __ORDER_BIG_ENDIAN__;
}

int32_t getValueOf___ORDER_LITTLE_ENDIAN__() {
    return __ORDER_LITTLE_ENDIAN__;
}

int32_t getValueOf___ORDER_PDP_ENDIAN__() {
    return __ORDER_PDP_ENDIAN__;
}

int32_t* tryGetValueOf___OpenBSD__(int32_t* value) {
#if defined(__OpenBSD__)
    *value = __OpenBSD__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___powerpc__(int32_t* value) {
#if defined(__powerpc__)
    *value = __powerpc__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___powerpc64__(int32_t* value) {
#if defined(__powerpc64__)
    *value = __powerpc64__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___riscv(int32_t* value) {
#if defined( __riscv)
    *value = __riscv;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___SH4__(int32_t* value) {
#if defined(__SH4__)
    *value = __SH4__;
#else
    value = NULL;
#endif
    return value;
}

int32_t getValueOf___SIZEOF_LONG__() {
    return __SIZEOF_LONG__;
}

int32_t getValueOf___SIZEOF_POINTER__() {
    return __SIZEOF_POINTER__;
}

int32_t* tryGetValueOf___s390__(int32_t* value) {
#if defined(__s390__)
    *value = __s390__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___s390x__(int32_t* value) {
#if defined(__s390x__)
    *value = __s390x__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___sh__(int32_t* value) {
#if defined(__sh__)
    *value = __sh__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___sparc64__(int32_t* value) {
#if defined(__sparc64__)
    *value = __sparc64__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___sparc__(int32_t* value) {
#if defined(__sparc__)
    *value = __sparc__;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___TIMESIZE(int32_t* value) {
#if defined(__TIMESIZE)
    *value = __TIMESIZE;
#else
    value = NULL;
#endif
    return value;
}

int32_t* tryGetValueOf___WORDSIZE(int32_t* value) {
#if defined (__WIN32__) || defined(__OpenBSD__)
#if !defined(__WORDSIZE)
    value = NULL;
#else
#error "__WORDSIZE defined"
#endif
#else
    *value = __WORDSIZE;
#endif
    return value;
}

int32_t* tryGetValueOf___x86_64__(int32_t* value) {
#if defined(__x86_64__)
    *value = __x86_64__;
#else
    value = NULL;
#endif
    return value;
}